src/gf.c

```diff
@@ -1730,6 +1730,78 @@ DLLEXPORT jl_value_t *jl_gf_invoke_lookup(jl_function_t *gf, jl_datatype_t *type
     return (jl_value_t*)m;
 }
 
+jl_function_t*
+jl_gf_invoke_get_specialization(jl_function_t *gf, jl_tupletype_t *types,
```

Comment:
I've thought about just convert the argument to `jl_gf_invoke` to tupletype and use this function to get the specialized function. I didn't bother too much (not that I like this kind of code duplication either though) because IMHO this is basically the same duplication of `jl_method_table_assoc_exact` and `jl_method_table_assoc_exact_by_type`.

Another way I've thought about is to just use a (more C friendly) (pointer to) array of types but again that kind of mean rewrite/merge `jl_method_table_assoc_exact` and `jl_method_table_assoc_exact_by_type` and I hoped to make this patch as small as possible.

Which one do you prefer?